### ISSUE TITLE
CopyObject: passing x-amz-copy-source-if-match and x-amz-copy-source-if-none-match headers not working as expected

### ISSUE THREAD

**Issue by vadmeste** [minio/minio#2851]
I tested the code below against Minio:

In Minio:
CopyObject doesn't work when SetMatchETag is set with correct or wrong ETag
CopyObject always works when SetMatchETagExcept is set, even when the actual etag is passed.  

Both work as expected with AWS S3.

```
package main

import (
    "log"
    "github.com/minio/minio-go"
)

func main() {
    s3Client, err := minio.New("localhost:9000", "JOTP3UO0V73PRNY0O4NI", "iBv5gq9MKNqNL1G/oIYpy+K2vGsnPATyMVWneTwg", false)
    if err != nil {
        log.Fatalln(err)
    }

    var copyConds = minio.NewCopyConditions()
    copyConds.SetMatchETag("bd891862ea3e22c93ed53a098218791d")
    // copyConds.SetMatchETagExcept("bd891862ea3e22c93ed53a098218791d")

    // Initiate copy object.
    err = s3Client.CopyObject("vadmeste", "copyobjectname", "/vadmeste/myobject", copyConds)
    if err != nil {
        log.Fatalln(err)
    }
    log.Println("Success.")
}
```

**Comment by vadmeste** [comment]
This is because I tested against FS which doesn't store ETags.

**Comment by harshavardhana** [comment]
> This is because I tested against FS which doesn't store ETags.

Well then its not a bug.

**Comment by vadmeste** [comment]
> Well then its not a bug.

Well, for the user, the behavior is just unpredictable especially if the api returns successful operation when it shouldn't.. If we can't return customized error here, I suggest to always return error (conditions not satisfied)

**Comment by harshavardhana** [comment]
> Well, for the user, the behavior is just unpredictable especially if the api returns successful operation when it shouldn't.. If we can't return customized error here, I suggest to always return error (conditions not satisfied)

Yes but we need to comply with HTTP spec behavior we can't return custom errors here.  Can you investigate HTTP spec behavior when we don't know ETag ?

**Comment by vadmeste** [comment]
> Yes but we need to comply with HTTP spec behavior we can't return custom errors here. Can you investigate HTTP spec behavior when we don't know ETag ?

First, I suggest to disable returning ETags in FS mode. So it won't make sense for clients to set if-match and if-none-match headers with any value other than "*"

This is a great resource but not clear what to do in our case:
https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/http-headers-status/http-headers-status%20v3%20draft.png

RFC either doesn't clearly say what to do, but probably we can conclude from this paragraph:
https://tools.ietf.org/html/rfc7232

```
 An origin server that receives an If-Match header field MUST evaluate
   the condition prior to performing the method (Section 5).  If the
   field-value is "*", the condition is false if the origin server does
   not have a current representation for the target resource.  If the
   field-value is a list of entity-tags, the condition is false if none
   of the listed tags match the entity-tag of the selected
   representation.
```

**Comment by harshavardhana** [comment]
We decided to succeed requests in case of FS for "ETag" if its not available.  We do the same with GetObject() as well.

**Comment by vadmeste** [comment]
> We decided to succeed requests in case of FS for "ETag" if its not available.

Ok.

> We do the same with GetObject() as well.

In FS, GetObject() already succeed when ETag is not available.

**Comment by harshavardhana** [comment]
Fixed by #2888
